File: 1D_DP/LC_5.py
# Written by: Jian-Yuan Lin
# Date: 2022.12.06

class Solution:
    def longestPalindrome(self, s: str) -> str:
        # begin from the middle and expand
        ans = ""
        ans_len = 0
        n = len(s)
        for i in range(n):
            # odd length palindrome
            l, r = i, i
            while l >= 0 and r < n and s[l] == s[r]:
                if r-l+1 > ans_len:
                    ans = s[l:r+1]
                    ans_len = r-l+1
                l -= 1
                r += 1
            # even length palindrome
            l, r = i, i+1
            while l >= 0 and r < n and s[l] == s[r]:
                if r-l+1 > ans_len:
                    ans = s[l:r+1]
                    ans_len = r-l+1
                l -= 1
                r += 1
        return ans

File: 1D_DP/LC_5.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.12.06

class Solution {
public:
    string longestPalindrome(string s) {
        // begin from the middle and expand
        int n = s.size();
        string ans;
        int ans_len = 0;
        for (int i = 0; i < n; i++){
            // odd length Palindrome
            int l = i, r = i;
            while (l >= 0 && r < n && s[l] == s[r]){
                if (r - l + 1 > ans_len){
                    ans_len = r - l + 1;
                    ans = s.substr(l, ans_len);
                }
                l--;
                r++;
            }
            // even length Palindrome
            l = i;
            r = i+1;
            while (l >= 0 && r < n && s[l] == s[r]){
                if (r - l + 1 > ans_len){
                    ans_len = r - l + 1;
                    ans = s.substr(l, ans_len);
                }
                l--;
                r++;
            }
        }
        
        return ans;
    }
};